### PR TITLE
bf: S3C-3303 empty notif config ok

### PR DESCRIPTION
**Pull request by dora-korpar**
Allow empty notification configuration as a means to delete existing notification configuration.

### CODE REVIEW

File: lib/models/NotificationConfiguration.js [hunk 1]
```diff
@@ -289,21 +289,23 @@ class NotificationConfiguration {
...
-                filterRules.forEach(f => {
-                    assert.strictEqual(typeof f.name, 'string');
-                    assert.strictEqual(typeof f.value, 'string');
-                });
-            }
-            assert.strictEqual(typeof id, 'string');
-        });
+        if (config.queueConfig) {
+            assert.strictEqual(Array.isArray(config.queueConfig), true);
+            config.queueConfig.forEach(q => {
```
Comment (miniscruff) [inline lib/models/NotificationConfiguration.js]: Not sure this matters, but I just want to compare two assertions and mention that even without asserting something you are effectively writing assertions in certain cases.
These two lines contain two asserts first the strict equal and second that queueConfig contains a forEach function.
The reason for this is that assert just takes some values or conditions and throws an exception if some condition is not met. Similarly trying to call a function on a variable that has no function of that name will also throw an exception because that condition was not met.

I also want to compare the output of these two assertions as sometimes not asserting can give more information in logs than writing an assert.

The first creates an exception that basically says `assertion error: false !== true` if the queue config is not an array. A message that is not too helpful without the line number that gives us the assertion.
The second "hidden" assert gives an error similar to `Uncaught TypeError: config.queueConfig.forEach is not a function`. In my opinion, the second error gives more valuable debug information without needing extra verbose assert messages. And because any error results in an invalid configuration it should work identically.

tl;dr the assert is not required as the forEach method is only found on arrays anyway.

File: lib/models/NotificationConfiguration.js [hunk 2]
```diff
@@ -289,21 +289,23 @@ class NotificationConfiguration {
...
-            if (filterRules) {
-                assert.strictEqual(Array.isArray(filterRules), true);
-                filterRules.forEach(f => {
-                    assert.strictEqual(typeof f.name, 'string');
-                    assert.strictEqual(typeof f.value, 'string');
-                });
-            }
-            assert.strictEqual(typeof id, 'string');
-        });
+        if (config.queueConfig) {
```
Comment (miniscruff) [inline lib/models/NotificationConfiguration.js]: Regardless of my other comment, I would turn this into a guard clause and early return over the extra tab.

File: lib/models/NotificationConfiguration.js (hunk 1 above)
Comment (dora-korpar) [reply lib/models/NotificationConfiguration.js]: i removed all the array asserts before `.forEach` uses

File: lib/models/NotificationConfiguration.js (hunk 2 above)
Comment (dora-korpar) [reply lib/models/NotificationConfiguration.js]: the way that function is used doesn't allow for it to have a return, so there's not really a way to make it a guard clause

**Comment by dora-korpar** [comment]
reset

File: lib/models/NotificationConfiguration.js (hunk 2 above)
Comment (miniscruff) [reply lib/models/NotificationConfiguration.js]: Does this not work?
```javascript
if (!config.queueConfig) {
    return
}
```

Comment (dora-korpar) [reply lib/models/NotificationConfiguration.js]: i mean yeah, but it's extra lines vs an extra tab. i prefer an extra tab. we can leave it up to jonathan if he wants to enforce a style

Comment (jonathan-gramain) [reply lib/models/NotificationConfiguration.js]: Personally I tend to prefer `return` for error cases to return early from the whole processing, and use a block when the flow logic becomes more complex. But I don't really mind in the end as long as the indentation level does not go up too much.

File: lib/models/NotificationConfiguration.js [hunk 3]
```diff
@@ -91,8 +91,8 @@ class NotificationConfiguration {
         }
         const queueConfig = notificationConfig.QueueConfiguration;
         if (!queueConfig || !queueConfig[0] || queueConfig[0] === '') {
```
Comment (jonathan-gramain) [inline lib/models/NotificationConfiguration.js]: Not changed in this PR, but the last test is superfluous (covered by the 2nd, as an empty string is false-ish)

File: lib/models/NotificationConfiguration.js [hunk 4]
```diff
@@ -254,7 +254,7 @@ class NotificationConfiguration {
      */
     static getConfigXML(config) {
         const xmlArray = [];
-        if (config) {
+        if (config && Object.keys(config).length > 0) {
```
Comment (jonathan-gramain) [inline lib/models/NotificationConfiguration.js]: Why not checking instead `if (config && config.queueConfig)`?

Comment (dora-korpar) [reply lib/models/NotificationConfiguration.js]: cos I copied the logic over from cloudserver where it made more sense. good catch

File: lib/models/NotificationConfiguration.js (hunk 2 above)
Comment (miniscruff) [reply lib/models/NotificationConfiguration.js]: For me when writing javascript, especially when using callbacks, the reduced indentation levels is usually preferred over the one saved line. The opening and close brace are required regardless, it is just the return line that is added with guard clauses.

**Review by miniscruff — APPROVED** [review]

**Comment by dora-korpar** [comment]
@bert-e reset

File: lib/models/NotificationConfiguration.js [hunk 5]
```diff
@@ -289,21 +289,20 @@ class NotificationConfiguration {
      */
     static validateConfig(config) {
         assert.strictEqual(typeof config, 'object');
-        assert.strictEqual(Array.isArray(config.queueConfig), true);
+        if (!config.queueConfig) return;
```
Comment (jonathan-gramain) [inline lib/models/NotificationConfiguration.js]: Can you add `{` and `}` and return on new line to conform to style guidelines

**Comment by dora-korpar** [comment]
@bert-e reset

File: lib/models/NotificationConfiguration.js [hunk 6]
```diff
@@ -289,21 +289,22 @@ class NotificationConfiguration {
...
             if (filterRules) {
-                assert.strictEqual(Array.isArray(filterRules), true);
                 filterRules.forEach(f => {
                     assert.strictEqual(typeof f.name, 'string');
                     assert.strictEqual(typeof f.value, 'string');
                 });
             }
             assert.strictEqual(typeof id, 'string');
         });
+        return;
```
Comment (jonathan-gramain) [inline lib/models/NotificationConfiguration.js:307]: I'm surprised this is needed, if it's linter complaint, fair enough

**Review by jonathan-gramain — APPROVED** [review]

File: lib/models/NotificationConfiguration.js (hunk 6 above)
Comment (dora-korpar) [reply lib/models/NotificationConfiguration.js:307]: yeah, just for the linter

**Comment by dora-korpar** [comment]
@bert-e approve
